File: python/GenerateSSCorr.py
import numpy as np

def GenerateLinearSSN(LocExp):

    # determine size of input data
    NumNode, NumSamp = LocExp.shape

    # create filter to extract upper diagonal since these are all symmetric matrices / saves memory; 
    #NodeIdx are the indices of the elements returned
    uF = np.triu(np.ones((NumNode, NumNode), dtype=int), 1)
    i_idx, j_idx = np.where(uF)
    NodeIdx = np.vstack([i_idx, j_idx]).T
    uF_vec = uF.T.flatten()

    #pre-computation step for LIONESS / SSN / SWEET
    AgNet_full = np.corrcoef(LocExp)
    AgNet = AgNet_full.flatten()[list(map(bool,uF_vec))]

    #pre-computation step for BONOBO (simplified for large N)
    AgCov = np.cov(LocExp)

    #additional pre-computation step for SWEET
    PCCmat = np.corrcoef(LocExp, rowvar=False)  # MATLAB corr(LocExp) treats columns as variables
    muPCC = (np.sum(PCCmat, axis=0) - 1) / (NumSamp - 1)
    minPCC = np.min(muPCC)
    maxPCC = np.max(muPCC)

    # allocate matrices
    #matrices to store the computed networks
    LION  = np.zeros((len(AgNet), NumSamp))
    SSN   = np.zeros((len(AgNet), NumSamp))
    SWEET = np.zeros((len(AgNet), NumSamp))
    BONO  = np.zeros((len(AgNet), NumSamp))

    
    for idx in range(NumSamp):

        #vector to extract all but sample idx
        idxvec = np.concatenate([np.arange(idx), np.arange(idx+1, NumSamp)])

        #compute Pearson correlation with all but sample idx; used for LIONESS / SSN / SWEET
        LocNet_full = np.corrcoef(LocExp[:, idxvec])
        LocNet = LocNet_full.flatten()[list(map(bool,uF_vec))]
#
        #LIONESS
        LION[:, idx] = NumSamp * (AgNet - LocNet) + LocNet
#
        #SSN
        SSN[:, idx] = (NumSamp - 2) * (AgNet - LocNet) / (1 - LocNet * LocNet)
#
        #SWEET
        Sp = (muPCC[idx] - minPCC + 0.01) / (maxPCC - minPCC + 0.01)
        SWEET[:, idx] = Sp * 0.1 * (NumSamp - 1) * (AgNet - LocNet) + LocNet
#
        #BONOBO
        LocCov = np.cov(LocExp[:, idxvec]) #BONOBO: covariance_matrix
        cov_diag = np.diag(LocCov)
        delta = 1 / (3 + 2 * np.mean(np.sqrt(cov_diag)) / np.var(cov_diag,ddof=1)) #BONOBO: delta

        sscov = delta * NumSamp * (AgCov - LocCov) + LocCov

        # sample-specific SD matrix
        sds = np.zeros((NumNode, NumNode))
        sds[np.diag_indices(NumNode)] = sscov.diagonal() ** (-0.5) #sample-specific standard deviations

        BON = sds @ sscov @ sds
        BONO[:, idx] = BON.flatten()[list(map(bool,uF_vec))]

    AllSSN = [LION, SSN, SWEET, BONO]
    return AllSSN, AgNet, NodeIdx

File: Supplemental/NetworkCode/GenerateSSCorr.m
function [AllSSN, AgNet, NodeIdx]=GenerateLinearSSN(LocExp);

% determine size of input data
[NumNode,NumSamp]=size(LocExp); 

% create filter to extract upper diagonal since these are all symmetric matrices / saves memory; NodeIdx are the indices of the elements returned
uF=triu(ones(NumNode),1); [i,j]=ind2sub([NumNode,NumNode], find(uF)); NodeIdx=[i,j];
uF=uF(:)==1;

% pre-computation step for LIONESS / SSN / SWEET
AgNet=corr(LocExp'); AgNet=AgNet(uF);

% additional pre-computation step for SWEET
PCCmat=corr(LocExp); muPCC=(sum(PCCmat)-1)/(NumSamp-1);
minPCC=min(muPCC); maxPCC=max(muPCC);

% matrices to store the computed networks
LION=zeros(length(AgNet),NumSamp);
SSN=zeros(length(AgNet),NumSamp);
SWEET=zeros(length(AgNet),NumSamp);
BONO=zeros(length(AgNet),NumSamp);

for(idx=1:NumSamp);
	% vector to extract all but sample idx
	idxvec=[1:(idx-1),(idx+1):NumSamp];

	% compute Pearson correlation with all but sample idx; used for LIONESS / SSN / SWEET
	LocNet=corr(LocExp(:,idxvec)'); LocNet=LocNet(uF);

	% LIONESS
	LION(:,idx)=(NumSamp)*(AgNet-LocNet)+LocNet; % wq=1/N

	% SSN
	SSN(:,idx)=(NumSamp-2)*(AgNet-LocNet)./(1-LocNet.*LocNet);

	% SWEET
	% compute Pearson correlation with all samples with sample idx duplicated; used for SWEET
	LocNet=corr([LocExp, LocExp(:,idx)]'); LocNet=LocNet(uF);
	Sp=(muPCC(idx)-minPCC+0.01)/(maxPCC-minPCC+0.01); % x=0.01
	SWEET(:,idx)=Sp*0.1*NumSamp*(LocNet-AgNet)+AgNet; % K=0.1

	% BONOBO
	cov_mat=cov(LocExp(:,idxvec)'); % BONOBO: covariance_matrix
	cov_diag=cov_mat(1:(NumNode+1):end);
	delta=1/(3+2*mean(sqrt(cov_diag))/var(cov_diag)); % BONOBO: delta
	A=LocExp(:,idx)-mean(LocExp,2);
	sscov=delta*A*A'+(1-delta)*cov_mat; % BONOBO: sscov
	sscov_diag=sqrt(sscov(1:(NumNode+1):end)); sscov_diag(sscov_diag==0)=1; % pull out the diagonal of sscov, set any zeros equal to 1
	sscov_eye=eye(NumNode); sscov_eye(1:(NumNode+1):end)=sscov_diag; % turn this into a diagonal matrix; BONOBO: diag
	sds=inv(sscov_eye);
	BON=sds*sscov*sds;
	BONO(:,idx)=BON(uF);
end
% merge matrices into one structure to return
AllSSN={LION, SSN, SWEET, BONO};

